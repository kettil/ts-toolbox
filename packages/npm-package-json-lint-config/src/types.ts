type DefaultObject = {
  rules: Record<string, readonly ['error' | 'off' | 'warning', unknown] | 'error' | 'off' | 'warning'>;
};

export type { DefaultObject };
