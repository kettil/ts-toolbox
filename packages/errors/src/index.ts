export { customAggregateError } from './lib/customAggregateError';
export { customError } from './lib/customError';
export { defaultErrors, defaultAggregateErrors } from './lib/defaultErrors';
export { isCustomAggregateError } from './lib/isCustomAggregateError';
export { isCustomError } from './lib/isCustomError';
export { isError } from './lib/isError';
