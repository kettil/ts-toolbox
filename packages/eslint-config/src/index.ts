import { createEslintConfig } from './configs/eslint';
import { createEslintCommentsConfig } from './configs/eslintComments';
import { createFormatjsConfig } from './configs/formatjs';
import { createImportConfig } from './configs/import';
import { createJestConfig } from './configs/jest';
import { createJsxA11yConfig } from './configs/jsxA11y';
import { createNodeConfig } from './configs/node';
import { createReactConfig } from './configs/react';
import { createReactHooksConfig } from './configs/reactHooks';
import { createRegexpConfig } from './configs/regexp';
import { createSonarjsConfig } from './configs/sonarjs';
import { createTestingLibraryConfig } from './configs/testingLibrary';
import { createTestingLibraryJestDomConfig } from './configs/testingLibraryJestDom';
import { createTypescriptConfig } from './configs/typescript';
import { createUnicornConfig } from './configs/unicorn';
import { deepMerge } from './lib/deepEqual';
import { defaultObject } from './lib/defaultObject';
import type { DefaultObject } from './types/defaultObject';

const createConfig = (): Readonly<DefaultObject> => {
  const plugins: readonly DefaultObject[] = [
    createEslintConfig(),
    createNodeConfig(),
    createTypescriptConfig(),

    // common
    createFormatjsConfig(),
    createImportConfig(),
    createUnicornConfig(),
    createEslintCommentsConfig(),
    createSonarjsConfig(),
    createRegexpConfig(),

    // react
    createReactConfig(),
    createReactHooksConfig(),
    createJsxA11yConfig(),

    // testing
    createJestConfig(),
    createTestingLibraryConfig(),
    createTestingLibraryJestDomConfig(),
  ];

  return plugins.reduce<DefaultObject>(
    (previousConfig, plugin) => ({
      env: deepMerge(previousConfig.env, plugin.env, ['env']),
      rules: { ...previousConfig.rules, ...plugin.rules },
      globals: deepMerge(previousConfig.globals, plugin.globals, ['globals']),
      settings: deepMerge(previousConfig.settings, plugin.settings, ['settings']),
      parserOptions: deepMerge(previousConfig.parserOptions, plugin.parserOptions, ['parserOptions']),
      parser: deepMerge(previousConfig.parser, plugin.parser, ['parser']),
      plugins: deepMerge(previousConfig.plugins, plugin.plugins, ['plugins']),
      overrides: deepMerge(previousConfig.overrides, plugin.overrides, ['overrides']),
    }),
    defaultObject,
  );
};

export { createConfig };
