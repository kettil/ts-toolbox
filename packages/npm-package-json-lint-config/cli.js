const { config } = require('./build/cli');

module.exports = config;
