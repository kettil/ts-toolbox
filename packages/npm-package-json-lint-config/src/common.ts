/* eslint-disable @typescript-eslint/naming-convention -- external definition */
import type { DefaultObject } from './types';

const config: DefaultObject = {
  rules: {
    // require node rules
    'require-author': 'off',
    'require-bin': 'off',
    'require-bugs': 'off',
    'require-bundledDependencies': 'off',
    'require-config': 'off',
    'require-contributors': 'error',
    'require-cpu': 'off',
    'require-dependencies': 'error',
    'require-description': 'error',
    'require-devDependencies': 'error',
    'require-directories': 'off',
    'require-engines': 'error',
    'require-files': 'off',
    'require-funding': 'off',
    'require-homepage': 'off',
    'require-keywords': 'error',
    'require-license': 'error',
    'require-main': 'off',
    'require-man': 'off',
    'require-module': 'off',
    'require-name': 'error',
    'require-optionalDependencies': 'off',
    'require-os': 'off',
    'require-peerDependencies': 'off',
    'require-preferGlobal': 'off',
    'require-private': 'error',
    'require-publishConfig': 'error',
    'require-repository-directory': 'off',
    'require-repository': 'error',
    'require-scripts': 'error',
    'require-types': 'off',
    'require-typings': 'off',
    'require-version': 'error',

    // type rules
    'bin-type': 'error',
    'bundledDependencies-type': 'error',
    'config-type': 'error',
    'cpu-type': 'error',
    'dependencies-type': 'error',
    'description-type': 'error',
    'devDependencies-type': 'error',
    'directories-type': 'error',
    'engines-type': 'error',
    'files-type': 'error',
    'homepage-type': 'error',
    'keywords-type': 'error',
    'license-type': 'error',
    'main-type': 'error',
    'man-type': 'error',
    'name-type': 'error',
    'optionalDependencies-type': 'error',
    'os-type': 'error',
    'peerDependencies-type': 'error',
    'preferGlobal-type': 'error',
    'private-type': 'error',
    'repository-type': 'error',
    'scripts-type': 'error',
    'version-type': 'error',

    // valid value rules
    'valid-values-author': 'off',
    'valid-values-engines': 'off',
    'valid-values-license': 'off',
    'valid-values-name-scope': 'off',
    'valid-values-private': 'off',
    'valid-values-publishConfig': 'off',

    // dependency rules
    'no-repeated-dependencies': 'error',

    // dependency rules - dependencies
    'no-absolute-version-dependencies': 'off',
    'no-archive-dependencies': 'error',
    'no-caret-version-dependencies': 'error',
    'no-file-dependencies': 'error',
    'no-git-dependencies': 'error',
    'no-restricted-dependencies': 'off',
    'no-restricted-pre-release-dependencies': 'off',
    'no-tilde-version-dependencies': 'error',
    'prefer-absolute-version-dependencies': 'error',
    'prefer-alphabetical-dependencies': 'error',
    'prefer-caret-version-dependencies': 'off',
    'prefer-no-version-zero-dependencies': 'off',
    'prefer-tilde-version-dependencies': 'off',

    // dependency rules - devDependencies
    'no-absolute-version-devDependencies': 'off',
    'no-archive-devDependencies': 'error',
    'no-caret-version-devDependencies': 'error',
    'no-file-devDependencies': 'error',
    'no-git-devDependencies': 'error',
    'no-restricted-devDependencies': 'off',
    'no-restricted-pre-release-devDependencies': 'off',
    'no-tilde-version-devDependencies': 'error',
    'prefer-absolute-version-devDependencies': 'error',
    'prefer-alphabetical-devDependencies': 'error',
    'prefer-caret-version-devDependencies': 'off',
    'prefer-no-version-zero-devDependencies': 'off',
    'prefer-tilde-version-devDependencies': 'off',

    // dependency rules - bundledDependencies
    'prefer-alphabetical-bundledDependencies': 'error',

    // dependency rules - optionalDependencies
    'prefer-alphabetical-optionalDependencies': 'error',

    // dependency rules - peerDependencies
    'prefer-alphabetical-peerDependencies': 'error',

    // scripts rules
    'prefer-alphabetical-scripts': 'error',
    'prefer-scripts': 'off',

    // format rules
    'description-format': [
      'error',
      {
        requireCapitalFirstLetter: true,
        requireEndingPeriod: true,
      },
    ],
    'name-format': 'error',
    'version-format': 'error',

    // package.json property rules
    'prefer-property-order': [
      'error',
      [
        'name',
        'version',
        'description',
        'license',
        'private',
        'contributors',
        'type',
        'main',
        'types',
        'module',
        'bin',
        'workspaces',
        'scripts',
        'dependencies',
        'devDependencies',
        'peerDependencies',
        'optionalDependencies',
        'homepage',
        'repository',
        'bugs',
        'keywords',
        'publishConfig',
        'engines',
      ],
    ],
    'no-duplicate-properties': 'error',

    // disallowed node rules
    'prefer-no-engineStrict': 'off',
    'prefer-no-dependencies': 'off',
    'prefer-no-devDependencies': 'off',
    'prefer-no-optionalDependencies': 'off',
    'prefer-no-peerDependencies': 'off',
  },
};

export { config };
