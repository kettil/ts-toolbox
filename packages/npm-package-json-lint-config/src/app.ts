/* eslint-disable @typescript-eslint/naming-convention -- external definition */
import { config as defaultConfig } from './common';
import type { DefaultObject } from './types';

const config: DefaultObject = {
  ...defaultConfig,
  rules: {
    ...defaultConfig.rules,

    'require-publishConfig': 'off',

    'repository-type': 'off',
    'homepage-type': 'off',
    'keywords-type': 'off',

    'valid-values-private': ['error', [true]],
  },
};

export { config };
