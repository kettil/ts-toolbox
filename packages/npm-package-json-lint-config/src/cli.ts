/* eslint-disable @typescript-eslint/naming-convention -- external definition */
import { config as defaultConfig } from './common';
import type { DefaultObject } from './types';

const config: DefaultObject = {
  ...defaultConfig,
  rules: {
    ...defaultConfig.rules,

    'require-bin': 'error',
  },
};

export { config };
