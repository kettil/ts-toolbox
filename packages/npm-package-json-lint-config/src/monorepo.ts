/* eslint-disable @typescript-eslint/naming-convention -- external definition */
import { config as defaultConfig } from './common';
import type { DefaultObject } from './types';

const config: DefaultObject = {
  ...defaultConfig,
  rules: {
    ...defaultConfig.rules,

    'require-dependencies': 'off',
    'require-homepage': 'off',
    'require-keywords': 'off',
    'require-publishConfig': 'off',
    'require-repository': 'off',

    'valid-values-private': ['error', [true]],
  },
};

export { config };
