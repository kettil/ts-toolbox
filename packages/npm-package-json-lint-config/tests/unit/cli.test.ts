import { config } from '../../src/cli';

describe('config - cli', () => {
  test('return config object', () => {
    expect(config).toMatchSnapshot();
  });
});
