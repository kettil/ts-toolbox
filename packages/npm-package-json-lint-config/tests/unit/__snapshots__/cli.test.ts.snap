// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config - cli return config object 1`] = `
Object {
  "rules": Object {
    "bin-type": "error",
    "bundledDependencies-type": "error",
    "config-type": "error",
    "cpu-type": "error",
    "dependencies-type": "error",
    "description-format": Array [
      "error",
      Object {
        "requireCapitalFirstLetter": true,
        "requireEndingPeriod": true,
      },
    ],
    "description-type": "error",
    "devDependencies-type": "error",
    "directories-type": "error",
    "engines-type": "error",
    "files-type": "error",
    "homepage-type": "error",
    "keywords-type": "error",
    "license-type": "error",
    "main-type": "error",
    "man-type": "error",
    "name-format": "error",
    "name-type": "error",
    "no-absolute-version-dependencies": "off",
    "no-absolute-version-devDependencies": "off",
    "no-archive-dependencies": "error",
    "no-archive-devDependencies": "error",
    "no-caret-version-dependencies": "error",
    "no-caret-version-devDependencies": "error",
    "no-duplicate-properties": "error",
    "no-file-dependencies": "error",
    "no-file-devDependencies": "error",
    "no-git-dependencies": "error",
    "no-git-devDependencies": "error",
    "no-repeated-dependencies": "error",
    "no-restricted-dependencies": "off",
    "no-restricted-devDependencies": "off",
    "no-restricted-pre-release-dependencies": "off",
    "no-restricted-pre-release-devDependencies": "off",
    "no-tilde-version-dependencies": "error",
    "no-tilde-version-devDependencies": "error",
    "optionalDependencies-type": "error",
    "os-type": "error",
    "peerDependencies-type": "error",
    "prefer-absolute-version-dependencies": "error",
    "prefer-absolute-version-devDependencies": "error",
    "prefer-alphabetical-bundledDependencies": "error",
    "prefer-alphabetical-dependencies": "error",
    "prefer-alphabetical-devDependencies": "error",
    "prefer-alphabetical-optionalDependencies": "error",
    "prefer-alphabetical-peerDependencies": "error",
    "prefer-alphabetical-scripts": "error",
    "prefer-caret-version-dependencies": "off",
    "prefer-caret-version-devDependencies": "off",
    "prefer-no-dependencies": "off",
    "prefer-no-devDependencies": "off",
    "prefer-no-engineStrict": "off",
    "prefer-no-optionalDependencies": "off",
    "prefer-no-peerDependencies": "off",
    "prefer-no-version-zero-dependencies": "off",
    "prefer-no-version-zero-devDependencies": "off",
    "prefer-property-order": Array [
      "error",
      Array [
        "name",
        "version",
        "description",
        "license",
        "private",
        "contributors",
        "type",
        "main",
        "types",
        "module",
        "bin",
        "workspaces",
        "scripts",
        "dependencies",
        "devDependencies",
        "peerDependencies",
        "optionalDependencies",
        "homepage",
        "repository",
        "bugs",
        "keywords",
        "publishConfig",
        "engines",
      ],
    ],
    "prefer-scripts": "off",
    "prefer-tilde-version-dependencies": "off",
    "prefer-tilde-version-devDependencies": "off",
    "preferGlobal-type": "error",
    "private-type": "error",
    "repository-type": "error",
    "require-author": "off",
    "require-bin": "error",
    "require-bugs": "off",
    "require-bundledDependencies": "off",
    "require-config": "off",
    "require-contributors": "error",
    "require-cpu": "off",
    "require-dependencies": "error",
    "require-description": "error",
    "require-devDependencies": "error",
    "require-directories": "off",
    "require-engines": "error",
    "require-files": "off",
    "require-funding": "off",
    "require-homepage": "off",
    "require-keywords": "error",
    "require-license": "error",
    "require-main": "off",
    "require-man": "off",
    "require-module": "off",
    "require-name": "error",
    "require-optionalDependencies": "off",
    "require-os": "off",
    "require-peerDependencies": "off",
    "require-preferGlobal": "off",
    "require-private": "error",
    "require-publishConfig": "error",
    "require-repository": "error",
    "require-repository-directory": "off",
    "require-scripts": "error",
    "require-types": "off",
    "require-typings": "off",
    "require-version": "error",
    "scripts-type": "error",
    "valid-values-author": "off",
    "valid-values-engines": "off",
    "valid-values-license": "off",
    "valid-values-name-scope": "off",
    "valid-values-private": "off",
    "valid-values-publishConfig": "off",
    "version-format": "error",
    "version-type": "error",
  },
}
`;
